File: ios/NitroFrameProcessor.swift
import Foundation
import NitroModules

@_silgen_name("CVIEBridgeActivateLicense")
private func CVIEBridgeActivateLicense(
  _ activationKey: UnsafePointer<CChar>?,
  _ deviceId: UnsafePointer<CChar>?
) -> Bool

@_silgen_name("CVIEBridgeCreate")
private func CVIEBridgeCreate(
  _ handleOut: UnsafeMutablePointer<UnsafeMutableRawPointer?>?
) -> Bool

@_silgen_name("CVIEBridgeDestroy")
private func CVIEBridgeDestroy(_ handle: UnsafeMutableRawPointer?)

@_silgen_name("CVIEBridgeConfigure")
private func CVIEBridgeConfigure(
  _ handle: UnsafeMutableRawPointer?,
  _ threads: Int32,
  _ parameterFilePath: UnsafePointer<CChar>?,
  _ width: Int32,
  _ height: Int32,
  _ setting: Int32
) -> Bool

@_silgen_name("CVIEBridgeEnhanceNextU8")
private func CVIEBridgeEnhanceNextU8(
  _ handle: UnsafeMutableRawPointer?,
  _ inputBytes: UnsafePointer<UInt8>?,
  _ outputBytes: UnsafeMutablePointer<UInt8>?,
  _ setting: Int32
) -> Bool

class NitroFrameProcessor: HybridNitroFrameProcessorSpec {
  private var isEnabled: Bool = true
  private var setting: Int32 = 0
  private var numThreads: Int32 = 1
  private var parameterFilePath: String = ""
  private var cvieHandle: UnsafeMutableRawPointer? = nil
  private var isLicenseActivated: Bool = false

  private var previousWidth: Int32 = -1
  private var previousHeight: Int32 = -1
  private var previousSetting: Int32 = -1
  private var previousParameterFilePath: String = ""

  public func setEnabled(value: Bool) throws {
    isEnabled = value
  }

  public func setNumThreads(numThreads: Double) throws {
    self.numThreads = Int32(max(1, Int(numThreads)))
  }

  public func setSetting(setting: Double) throws {
    self.setting = Int32(setting)
  }

  public func setParameterFilePath(path: String) throws {
    self.parameterFilePath = path
  }

  public func activateLicense(activationKey: String, deviceId: String) throws -> Bool {
    let activated = activationKey.withCString { activationKeyCString in
      deviceId.withCString { deviceIdCString in
        CVIEBridgeActivateLicense(activationKeyCString, deviceIdCString)
      }
    }
    isLicenseActivated = activated
    return activated
  }

  public func processFrame(
    width: Double,
    height: Double,
    input: ArrayBuffer
  ) throws -> ArrayBuffer {
    if !isEnabled { return input }

    let imageWidth = Int32(width)
    let imageHeight = Int32(height)
    if imageWidth <= 0 || imageHeight <= 0 { return input }
    if parameterFilePath.isEmpty { return input }

    do {
      try ensureInstanceInitialized(width: imageWidth, height: imageHeight)
      let inputData = input.toData(copyIfNeeded: true)
      var inputBytes = [UInt8](inputData)
      var outputBytes = [UInt8](repeating: 0, count: inputBytes.count)

      let result = CVIEBridgeEnhanceNextU8(
        cvieHandle,
        &inputBytes,
        &outputBytes,
        setting
      )
      if !result {
        return input
      }

      return try ArrayBuffer.copy(data: Data(outputBytes))
    } catch {
      return input
    }
  }

  private func ensureInstanceInitialized(width: Int32, height: Int32) throws {
    let needsRecreate =
      cvieHandle == nil ||
      previousWidth != width ||
      previousHeight != height ||
      previousSetting != setting ||
      previousParameterFilePath != parameterFilePath

    if !needsRecreate { return }

    if let handle = cvieHandle {
      CVIEBridgeDestroy(handle)
      cvieHandle = nil
    }

    var handle: UnsafeMutableRawPointer?
    guard CVIEBridgeCreate(&handle) else {
      throw NSError(domain: "NitroFrameProcessor", code: 1)
    }
    cvieHandle = handle

    guard isLicenseActivated else {
      throw NSError(domain: "NitroFrameProcessor", code: 2)
    }

    let configured = parameterFilePath.withCString { path in
      CVIEBridgeConfigure(
        cvieHandle,
        Int32(numThreads),
        path,
        width,
        height,
        setting
      )
    }
    if !configured {
      throw NSError(domain: "NitroFrameProcessor", code: 3)
    }

    previousWidth = width
    previousHeight = height
    previousSetting = setting
    previousParameterFilePath = parameterFilePath
  }
}

File: android/src/main/java/com/margelo/nitro/nitroframeprocessor/NitroFrameProcessor.kt
package com.margelo.nitro.nitroframeprocessor
  
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.ArrayBuffer
import java.nio.ByteBuffer
import java.util.concurrent.atomic.AtomicBoolean
import se.contextvision.cvie.Cvie
import se.contextvision.cvie.CvieDataFlag
import se.contextvision.cvie.CvieLibrary
import se.contextvision.cvie.Cvlm

@DoNotStrip
class NitroFrameProcessor : HybridNitroFrameProcessorSpec() {
  private var isEnabled = true
  private var setting = 0
  private var numThreads = 1
  private var parameterFilePath = ""
  private var cvieHandle: CvieLibrary.HCVIE? = null
  private val isLicenseActivated = AtomicBoolean(false)

  private var previousWidth = -1
  private var previousHeight = -1
  private var previousSetting = -1
  private var previousParameterFilePath = ""

  override fun setEnabled(value: Boolean) {
    isEnabled = value
  }

  override fun setNumThreads(numThreads: Double) {
    this.numThreads = numThreads.toInt().coerceAtLeast(1)
  }

  override fun setSetting(setting: Double) {
    this.setting = setting.toInt()
  }

  override fun setParameterFilePath(path: String) {
    this.parameterFilePath = path
  }

  override fun activateLicense(activationKey: String, deviceId: String): Boolean {
    return try {
      Cvlm.SetParameterValue(Cvlm.LM_INIT, deviceId)
      val modules = Cvlm.GetPossibleModules()
      for (index in modules.indices) {
        if (Cvlm.SetKey(index, activationKey)) {
          isLicenseActivated.set(true)
          return true
        }
      }
      false
    } catch (_: Throwable) {
      false
    }
  }

  override fun processFrame(width: Double, height: Double, input: ArrayBuffer): ArrayBuffer {
    if (!isEnabled) return input

    val imageWidth = width.toInt()
    val imageHeight = height.toInt()
    if (imageWidth <= 0 || imageHeight <= 0) return input
    if (parameterFilePath.isEmpty()) return input

    return try {
      ensureInstanceInitialized(imageWidth, imageHeight)
      val inBuffer = input.getBuffer(copyIfNeeded = true)
      val outBuffer = ByteBuffer.allocateDirect(inBuffer.capacity())
      Cvie.CVIEEnhanceNext(requireNotNull(cvieHandle), inBuffer, outBuffer, setting)
      outBuffer.rewind()
      ArrayBuffer.copy(outBuffer)
    } catch (_: Throwable) {
      input
    }
  }

  private fun ensureInstanceInitialized(width: Int, height: Int) {
    val needsRecreate =
      cvieHandle == null ||
        previousWidth != width ||
        previousHeight != height ||
        previousSetting != setting ||
        previousParameterFilePath != parameterFilePath

    if (!needsRecreate) return

    cvieHandle?.let { Cvie.CVIEDestroy(it) }
    cvieHandle = Cvie.CVIECreate(0)

    if (!isLicenseActivated.get()) {
      throw IllegalStateException("CVIE license is not activated.")
    }

    val handle = requireNotNull(cvieHandle)
    Cvie.CVIESetThreads(handle, numThreads)
    Cvie.CVIESetParameterFile(handle, parameterFilePath)
    Cvie.CVIEEnhanceSetup(handle, width, height, CvieDataFlag.U8.code, setting)

    previousWidth = width
    previousHeight = height
    previousSetting = setting
    previousParameterFilePath = parameterFilePath
  }

}
